godot-core/src/builtin/array.rs

```diff
@@ -185,61 +282,189 @@ impl Array {
         let end = end.min(len);
         let step = step.unwrap_or(1);
         assert!(step != 0);
-        self.as_inner()
-            .slice(to_i64(begin), to_i64(end), step.try_into().unwrap(), true)
+        let slice: Array =
+            self.as_inner()
+                .slice(to_i64(begin), to_i64(end), step.try_into().unwrap(), true);
+        // SAFETY: slice() returns a typed array with the same type as Self
+        unsafe { slice.assume_type() }
+    }
+
+    /// Appends another array at the end of this array. Equivalent of `append_array` in GDScript.
+    pub fn extend_array(&mut self, other: TypedArray<T>) {
+        // SAFETY: Read-only arrays are covariant: conversion to a variant array is fine as long as
+        // we don't insert values into it afterwards, and `append_array()` doesn't do that.
+        let other: Array = unsafe { other.assume_type() };
+        self.as_inner().append_array(other);
+    }
+
+    /// Returns the runtime type info of this array.
+    fn type_info(&self) -> TypeInfo {
+        let variant_type =
+            VariantType::from_sys(self.as_inner().get_typed_builtin().try_into().unwrap());
+        let class_name = self.as_inner().get_typed_class_name();
+        TypeInfo {
+            variant_type,
+            class_name,
+        }
+    }
+
+    /// Checks that the inner array has the correct type set on it for storing elements of type `T`.
+    ///
+    /// # Panics
+    ///
+    /// If the inner type doesn't match `T` or no type is set at all.
+    fn check_type(&self) {
+        assert_eq!(self.type_info(), TypeInfo::new::<T>());
+    }
+
+    /// Sets the type of the inner array. Can only be called once, directly after creation.
+    fn set_typed(&mut self) {
+        debug_assert!(self.is_empty());
+        debug_assert!(!self.type_info().is_typed());
+        let type_info = TypeInfo::new::<T>();
+        if type_info.is_typed() {
+            let script = Variant::nil();
+            unsafe {
+                (interface_fn!(array_set_typed))(
```

Comment:
The extra parentheses around `interface_fn!(array_set_typed)` should not be necessary (see existing invocations all over the codebase).